File: .agents/skills/bio-single-cell-multimodal-integration/examples/cite_seq_analysis.py
# Reference: numpy 1.26+, scanpy 1.10+ | Verify API if version differs
import muon as mu
import scanpy as sc
import anndata as ad
import numpy as np

# Read 10X multiome data (RNA + ADT)
# muon reads 10X filtered_feature_bc_matrix folder with both modalities
mdata = mu.read_10x_h5('filtered_feature_bc_matrix.h5')

# Access individual modalities
rna = mdata.mod['rna']
adt = mdata.mod['prot']  # ADT/protein

# RNA QC
rna.var['mt'] = rna.var_names.str.startswith('MT-')
sc.pp.calculate_qc_metrics(rna, qc_vars=['mt'], percent_top=None, inplace=True)

# Filter cells - keep cells present in both modalities
mu.pp.intersect_obs(mdata)

# Filter by QC metrics
# nFeature_RNA > 200: minimum gene complexity
# nFeature_RNA < 5000: remove potential doublets
# percent.mt < 20: remove dying cells
rna_mask = (rna.obs['n_genes_by_counts'] > 200) & (rna.obs['n_genes_by_counts'] < 5000) & (rna.obs['pct_counts_mt'] < 20)
mdata = mdata[rna_mask].copy()

# RNA preprocessing
sc.pp.normalize_total(rna, target_sum=1e4)
sc.pp.log1p(rna)
sc.pp.highly_variable_genes(rna, n_top_genes=2000)
sc.pp.scale(rna, max_value=10)
sc.tl.pca(rna, n_comps=30)

# ADT preprocessing (CLR normalization)
# margin=0: normalize across cells (standard for ADT)
mu.prot.pp.clr(adt, axis=0)
sc.pp.scale(adt, max_value=10)
# Use all ADT features for PCA (typically 10-200 markers)
sc.tl.pca(adt, n_comps=min(18, adt.n_vars - 1))

# Weighted nearest neighbors (WNN) for multimodal integration
mu.pp.neighbors(mdata, key_added='wnn')

# Clustering on WNN graph
sc.tl.leiden(mdata, resolution=0.5, key_added='wnn_clusters', neighbors_key='wnn')

# UMAP on WNN
sc.tl.umap(mdata, neighbors_key='wnn')

# Save UMAP plot
sc.settings.figdir = './'
sc.pl.umap(mdata, color='wnn_clusters', save='_cite_seq_wnn.pdf')

# Save processed data
mdata.write('cite_seq_analyzed.h5mu')
print('Analysis complete. Saved to cite_seq_analyzed.h5mu')

File: .agents/skills/bio-single-cell-multimodal-integration/examples/cite_seq_analysis.R
# Reference: numpy 1.26+, scanpy 1.10+ | Verify API if version differs
library(Seurat)

data <- Read10X('filtered_feature_bc_matrix/')
rna_counts <- data$`Gene Expression`
adt_counts <- data$`Antibody Capture`

obj <- CreateSeuratObject(counts = rna_counts, assay = 'RNA', min.cells = 3)
obj[['ADT']] <- CreateAssayObject(counts = adt_counts[, colnames(obj)])

obj <- PercentageFeatureSet(obj, pattern = '^MT-', col.name = 'percent.mt')
obj <- subset(obj, nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mt < 20)

obj <- NormalizeData(obj, assay = 'RNA')
obj <- FindVariableFeatures(obj, assay = 'RNA')
obj <- ScaleData(obj, assay = 'RNA')
obj <- RunPCA(obj, assay = 'RNA', reduction.name = 'pca')

obj <- NormalizeData(obj, assay = 'ADT', normalization.method = 'CLR', margin = 2)
obj <- ScaleData(obj, assay = 'ADT')
obj <- RunPCA(obj, assay = 'ADT', reduction.name = 'apca',
              features = rownames(obj[['ADT']]), npcs = min(18, nrow(obj[['ADT']]) - 1))

obj <- FindMultiModalNeighbors(obj, reduction.list = list('pca', 'apca'),
                                dims.list = list(1:30, 1:min(18, ncol(obj[['apca']]))))
obj <- FindClusters(obj, graph.name = 'wsnn', resolution = 0.5)
obj <- RunUMAP(obj, nn.name = 'weighted.nn', reduction.name = 'wnn.umap')

pdf('cite_seq_wnn_umap.pdf', width = 10, height = 8)
print(DimPlot(obj, reduction = 'wnn.umap', label = TRUE))
dev.off()

saveRDS(obj, 'cite_seq_analyzed.rds')
cat('Analysis complete. Saved to cite_seq_analyzed.rds\n')
